clients/drcachesim/tools/trace_filter.cpp

```diff
@@ -0,0 +1,217 @@
+            return false;
+        }
+
+        thread_data_.push_back(shard_data_t(static_cast<int>(thread_data_.size()),
+                                            std::move(reader), std::move(writer),
+                                            trace_path, output_path));
+    }
+    // Like raw2trace, we use a simple round-robin static work assigment.  This
+    // could be improved later with dynamic work queue for better load balancing.
+    if (worker_count_ <= 0)
+        worker_count_ = std::thread::hardware_concurrency();
+    worker_tasks_.resize(worker_count_);
+    int worker = 0;
+    for (size_t i = 0; i < thread_data_.size(); ++i) {
+        VPRINT(this, 2, "Worker %d assigned trace shard %zd\n", worker, i);
+        worker_tasks_[worker].push_back(&thread_data_[i]);
+        thread_data_[i].worker = worker;
+        worker = (worker + 1) % worker_count_;
+    }
+    trace_end_ = std::unique_ptr<default_trace_entry_file_reader_t>(
+        new default_trace_entry_file_reader_t());
+    return true;
+}
+
+void
+trace_filter_t::process_tasks(std::vector<shard_data_t *> *tasks)
+{
+    if (tasks->empty()) {
+        VPRINT(this, 1, "Worker has no tasks\n");
+        return;
+    }
+    VPRINT(this, 1, "Worker %d assigned %zd task(s)\n", (*tasks)[0]->worker,
+           tasks->size());
+    for (shard_data_t *tdata : *tasks) {
+        VPRINT(this, 1, "Worker %d starting on trace shard %d\n", tdata->worker,
+               tdata->index);
+        if (!tdata->iter->init()) {
+            tdata->error = "Failed to read from trace" + tdata->trace_file;
+            return;
+        }
+        for (; *tdata->iter != *trace_end_; ++(*tdata->iter)) {
+            const trace_entry_t &entry = **tdata->iter;
+            // TODO i#5675: Filter entries here.
+            if (!tdata->writer->write((char *)&entry, sizeof(entry))) {
+                tdata->error = "Failed to write to output file " + tdata->output_file;
+                return;
+            }
+        }
+        VPRINT(this, 1, "Worker %d finished trace shard %d\n", tdata->worker,
+               tdata->index);
+    }
+}
+
+bool
+trace_filter_t::run()
+{
+    if (error_string_ != "")
+        return false;
+    // XXX i#3286: Add a %-completed progress message by looking at the file sizes.
+    if (worker_count_ <= 0) {
+        error_string_ = "Invalid worker count: must be > 0";
+        return false;
+    }
+    std::vector<std::thread> threads;
+    VPRINT(this, 1, "Creating %d worker threads\n", worker_count_);
+    threads.reserve(worker_count_);
+    for (int i = 0; i < worker_count_; ++i) {
+        threads.emplace_back(
+            std::thread(&trace_filter_t::process_tasks, this, &worker_tasks_[i]));
+    }
+    for (std::thread &thread : threads)
+        thread.join();
+    for (auto &tdata : thread_data_) {
+        if (!tdata.error.empty()) {
+            error_string_ = tdata.error;
+            return false;
+        }
+    }
+    return true;
```

Comment:
Another function that is basically identical to the analyzer_t code if the serial mode if() is removed.  More evidence to try a subclass.

Reply:
Converted analyzer_t to a template in https://github.com/DynamoRIO/dynamorio/pull/5720.